ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm

```diff
@@ -0,0 +1,778 @@
+
+    # Check if the keyword is prefixed by:
+    #     -  a '-': in this case the corresponding line must be unconditionally 
+    #               commented out if it is present
+    #     -  a '*': in this case the corresponding line must be commented out if
+    #               it is present and the option is undefined
+    my $comment_line = 0;
+    if ( $keyword =~ /^-/ ) {
+      $keyword =~ s/^-//;
+      $comment_line = 1;
+    } elsif ( $keyword =~ /^\?/ ) {
+      $keyword =~ s/^\?//;
+      $remove_if_undef = 1;
+      $self->debug(2,"$function_name: 'remove_if_undef' option set for the current rule");
+    }
+
+    # Split different elements of the rule
+    ($rule, my $line_fmt, my $value_fmt) = split /;/, $rule;
+    unless ( $line_fmt ) {
+      $line_fmt = LINE_FORMAT_DEFAULT;
+    }
+    my $value_opt;
+    if ( $value_fmt ) {
+      ($value_fmt, $value_opt) = split /:/, $value_fmt;
+    }else {
+      $value_fmt = LINE_VALUE_AS_IS;
+    }
+    unless ( defined($value_opt) ) {
+      $value_opt = LINE_VALUE_OPT_NONE;      
+    }
+
+    (my $condition, my $tmp) = split /->/, $rule;
+    if ( $tmp ) {
+      $rule = $tmp;
+    } else {
+      $condition = "";
+    }
+    $self->debug(1,"$function_name: processing rule ".$rule_id."(variable=>>>".$keyword.
+                      "<<<, comment_line=".$comment_line.", condition=>>>".$condition."<<<, rule=>>>".$rule."<<<, fmt=".$line_fmt.")");
+
+    # Check if only rules with ALWAYS conditions must be applied 
+    if ( $parser_options->{always_rules_only} ) {
+      if ( $condition eq "ALWAYS" ) {
+        $condition = '';
+      } else {
+        $self->debug(1,"$function_name: rule ignored (ALWAYS condition not set)");
+        next;
+      }
+    }
+
+    # If the keyword was "negated", remove (comment out) configuration line if present and enabled
+    if ( $comment_line ) {
+      $self->debug(1,"$function_name: keyword '$keyword' negated, removing configuration line");
+      $self->removeConfigLine($fh,$keyword,$line_fmt);
+      next;
+    }
+
+    # Check if rule condition is met if one is defined
+    unless ( $condition eq "" ) {
+      $self->debug(1,"$function_name: checking condition >>>$condition<<<");
+
+      # Condition may be negated if it starts with a !
+      my $negate = 0;
+      if ( $condition =~ /^!/ ) {
+        $negate = 1;
+        $condition =~ s/^!//;
+      }
+      my ($cond_attribute,$cond_option_set) = split /:/, $condition;
+      unless ( $cond_option_set ) {
+        $cond_option_set = $cond_attribute;
+        $cond_attribute = "";
+      }
+      $self->debug(2,"$function_name: condition option set = '$cond_option_set', ".
+                         "condition attribute = '$cond_attribute', negate=$negate");
+      my $cond_satisfied = 1;
+      if ( $cond_attribute ) {
+        # Due to an exists() flaw, testing directly exists($config_options->{$cond_option_set}->{$cond_attribute}) will spring
```

Comment:
that's not a flaw in exists, it's perl autovivication.
it also means that you can loose the first `exists` and juts write
```perl
my $tmp = $config_options->{$cond_option_set} && exists($config_options->{$cond_option_set}->{$cond_attribute});
```
and use the `$tmp` if both expressions are indeed the same (and should be). it's very hard to compare/review copy/pasted code.